### PR TITLE
Make OSSL_LIB_CTX_load_config thread safe

### LINKED ISSUE #18226: OSSL_LIB_CTX_load_config is not thread safe
When OSSL_LIB_CTX_load_config is called simultaneously from multiple threads, the program abort with the following trace:

 ```
(gdb) where
#0  0x00007ffff7623387 in raise () from /lib64/libc.so.6
#1  0x00007ffff7624a78 in abort () from /lib64/libc.so.6
#2  0x00007ffff7665ed7 in __libc_message () from /lib64/libc.so.6
#3  0x00007ffff766c3e4 in malloc_printerr () from /lib64/libc.so.6
#4  0x00007ffff76716e0 in _int_realloc () from /lib64/libc.so.6
#5  0x00007ffff7672d82 in realloc () from /lib64/libc.so.6
#6  0x00000000004fa265 in CRYPTO_realloc (str=0x7ffff0006390, num=104, file=0x734de8 "crypto/stack/stack.c", line=210) at crypto/mem.c:214
#7  0x000000000052e1b8 in sk_reserve (st=0x7ffff0009570, n=1, exact=0) at crypto/stack/stack.c:210
#8  0x000000000052e2f4 in OPENSSL_sk_insert (st=0x7ffff0009570, data=0x7fffe8005ba0, loc=9) at crypto/stack/stack.c:254
#9  0x000000000052e79b in OPENSSL_sk_push (st=0x7ffff0009570, data=0x7fffe8005ba0) at crypto/stack/stack.c:368
#10 0x0000000000496625 in sk_CONF_IMODULE_push (sk=0x7ffff0009570, ptr=0x7fffe8005ba0) at crypto/conf/conf_mod.c:27
#11 0x00000000004970b6 in module_init (pmod=0x7ffff0002c10, name=0x7fffe80037e0 "providers", value=0x7fffe8003820 "provider_sect", cnf=0x7fffe8009770) at crypto/conf/conf_mod.c:390
#12 0x0000000000496b64 in module_run (cnf=0x7fffe8009770, name=0x7fffe80037e0 "providers", value=0x7fffe8003820 "provider_sect", flags=0) at crypto/conf/conf_mod.c:239
#13 0x0000000000496837 in CONF_modules_load (cnf=0x7fffe8009770, appname=0x0, flags=0) at crypto/conf/conf_mod.c:138
#14 0x00000000004969b0 in CONF_modules_load_file_ex (libctx=0x7fffe8009d00, filename=0x70ee50 "/tmp/openssl.cnf", appname=0x0, flags=0)
    at crypto/conf/conf_mod.c:181
#15 0x00000000004f52f8 in OSSL_LIB_CTX_load_config (ctx=0x7fffe8009d00, config_file=0x70ee50 "/tmp/openssl.cnf") at crypto/context.c:234
```
Each thread allocates its own library context and load config.
Please see the sample code below which reproduces the issue.  With mutex held while OSSL_LIB_CTX_load_config() is called, we don't see the issue.

```
#include <stdio.h>
#include <stdlib.h>
#include <openssl/bio.h>
#include <openssl/err.h>
#include <openssl/ssl.h>

pthread_mutex_t lock;

void *libctx_init(void *arg)
{
  int           rv;
  int           i;
  int           num_loops = (int)arg;
  OSSL_LIB_CTX *libctx = NULL;
  const char   *path = "/ade/mmiyashi_network_openssl/oracle/crypto/lib/openssl.cnf";

  for (i= 0; i < num_loops; i++)
  {
    printf("[%u][%d] LoadConfig called\n", pthread_self(), i);

    libctx = OSSL_LIB_CTX_new();
    if (libctx == NULL)
    {
       printf("[%u] Failed to allocate LIB_CTX\n", pthread_self());
    }
   
    //pthread_mutex_lock(&lock);
    rv = OSSL_LIB_CTX_load_config(libctx, path);
    //pthread_mutex_unlock(&lock);
    if (rv != 1)
    {
       printf("[%u] Failed to load config\n", pthread_self());
    }

    printf("[%u][%d] LoadConfig successful\n", pthread_self(), i);
  }
}

int main(int argc, char *argv[])
{
  int        rv;
  int        i;
  int        num_loops = 1;
  int        num_threads = 5;
  pthread_t  threads[1024];

  if (argc >= 2)
  {
     num_threads = atoi(argv[1]);
     if (num_threads > 1024)
     {
         printf("num_threads[%d] too largs. Reset to 1024\n", num_threads);
         num_threads = 1024;
     }
  }
  if (argc >= 3)
  {
     num_loops = atoi(argv[2]);
  }
  if (pthread_mutex_init(&lock, NULL) != 0) {
    printf("mutex init has failed\n");
    return 1;
  }

  SSL_load_error_strings();
  SSL_library_init();

  fprintf(stderr, "Initialize libctx\n");
  for (i = 0; i < num_threads; i++)
  {
    rv = pthread_create(&threads[i], NULL, libctx_init, (void *)num_loops);
    if (rv != 0) {
      printf("[%d]: pthread_create filed\n", i);
    }
  }

  for (i = 0; i < num_threads; i++)
  {
    pthread_join(threads[i], NULL);
  }

  return 0;
}
```

Is this an multi-threaded issue in OpenSSL? or an issue in the test code.

The issue is seen with OpenSSL 3.0.0.

Thank you,

-- misaki

### PR DESCRIPTION
**Pull request by hlandau**
Fixes #18226.

### CODE REVIEW

**Review by mattcaswell — DISMISSED** [review]

**Comment by hlandau** [comment]
Updated.

**Comment by hlandau** [comment]
Updated.

File: crypto/conf/conf_mod.c [hunk 1]
```diff
@@ -328,35 +328,33 @@ static CONF_MODULE *module_add(DSO *dso, const char *name,
...
-    }
+    if (!sk_CONF_MODULE_push(supported_modules, tmod))
+        goto err;
 
     CRYPTO_THREAD_unlock(module_list_lock);
     return tmod;
+
+ err:
+    CRYPTO_THREAD_unlock(module_list_lock);
+    if (tmod) {
```
Comment (t8m) [inline crypto/conf/conf_mod.c:353]: `if (tmod != NULL) {`

File: crypto/conf/conf_mod.c [hunk 2]
```diff
@@ -423,7 +475,11 @@ void CONF_modules_unload(int all)
 {
     int i;
     CONF_MODULE *md;
-    CONF_modules_finish();
+    CONF_modules_finish(); /* also inits module list lock */
```
Comment (paulidale) [inline crypto/conf/conf_mod.c:478]: blank line above :)

**Review by paulidale — APPROVED** [review]
With the nits fixed.

**Comment by hlandau** [comment]
Updated.

**Review by t8m — APPROVED** [review]

**Comment by beldmit** [comment]
Don't we need tests (e.g. from #18226) to be added to this PR?

**Comment by t8m** [comment]
@hlandau Would you please consider adding the test from #18226 ?

Perhaps as a separate PR as this one is done?

**Comment by hlandau** [comment]
@t8m Not sure I follow. This is a race condition resulting in a crash, there's no simple test for it. There's no guarantee that it will be reproducible on any given architecture. So I don't see any reliable way to test for  this.

**Comment by hlandau** [comment]
Added a basic test to threadstest, suggestions welcome.

File: crypto/conf/conf_mod.c [hunk 3]
```diff
@@ -379,21 +423,29 @@ static int module_init(CONF_MODULE *pmod, const char *name, const char *value,
...
+        goto err;
+
+    if (!CRYPTO_THREAD_write_lock(module_list_lock))
+        goto err;
+
     if (initialized_modules == NULL) {
         initialized_modules = sk_CONF_IMODULE_new_null();
-        if (!initialized_modules) {
+        if (initialized_modules == NULL) {
             ERR_raise(ERR_LIB_CONF, ERR_R_MALLOC_FAILURE);
```
Comment (mattcaswell) [inline crypto/conf/conf_mod.c:435]: Looks like we fail to unlock `module_list_lock` here

File: test/threadstest.c [hunk 4]
```diff
@@ -654,6 +654,19 @@ static int test_obj_add(void)
...
+        multi_success = 0;
+}
+
+static int test_lib_ctx_load_config(void)
+{
+    return thread_run_test(&test_lib_ctx_load_config_worker,
+                           MAXIMUM_THREADS, &test_lib_ctx_load_config_worker,
+                           1, default_provider);
+}
+
```
Comment (mattcaswell) [inline test/threadstest.c:669]: It would be interesting to know whether this test without the fix in this PR would have caught the problem.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: test/threadstest.c (hunk 4 above)
Comment (hlandau) [reply test/threadstest.c:669]: https://github.com/openssl/openssl/pull/18405

Caught! Thank you tsan.

**Comment by hlandau** [comment]
Updated, test is confirmed to catch this, ready for review.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
@hlandau can you please squash the fixup commits and force-push? That won't require re-approval if there are no other changes. The reason is that I'd have to manually merge this as the initial commits conflict with the current master so a rebase is not possible.

**Comment by t8m** [comment]
ping @hlandau

**Comment by t8m** [comment]
Merged to master and 3.0 branches (after fixing a trivial conflict).

**Pull request closed without merging**